Hashes/SHA1.js

```diff
@@ -0,0 +1,183 @@
+ *
+ * @param {string} str - string represention (binary/hex)
+ * @param {int} bits - total number of bits wanted
+ * @return {string} - string represention padding with empty (0) bits
+ *
+ * @example
+ *      pad("10011", 8); // "00010011"
+ */
+function pad(str, bits) {
+    let res = str;
+    while (res.length % bits !== 0) {
+        res = "0" + res;
+    }
+    return res;
+}
+
+/**
+ * Separates string into chunks of the same size
+ *
+ * @param {string} str - string to separate into chunks
+ * @param {int} size - number of characters wanted in each chunk
+ * @return {array} - array of original string split into chunks
+ *
+ * @example
+ *      chunkify("this is a test", 2); // ["th", "is", " i", "s ", "a ", "te", "st"]
+ */
+function chunkify(str, size) {
+    let chunks = [];
+    for (let i = 0; i < str.length; i += size) {
+        chunks.push(str.slice(i, i + size));
+    }
+    return chunks;
+}
+
+/**
+ * Rotates string represention of bits to th left
+ *
+ * @param {string} bits - string representation of bits
+ * @param {int} turns - number of rotations to make
+ * @return {string} - string representation of bits after rotation
+ *
+ * @example
+ *      rotateLeft("1011", 3); // "1101"
+ */
+function rotateLeft(bits, turns) {
+    return bits.substr(turns) + bits.substr(0, turns);
+}
+
+/**
+ * Pre-processes message to feed the algorithm loop
+ *
+ * @param {string} message - message to pre-process
+ * @return {string} - processed message
+ */
+function preProcess(message) {
+    //covert message to binary representation padded to
+    //8 bits, and add 1
+    let m = message.split("")
+        .map(e => e.charCodeAt(0))
+        .map(e => e.toString(2))
+        .map(e => pad(e, 8))
+        .join("") + "1";
+
+    //extend message by adding empty bits (0)
+    while (m.length % 512 !== 448) {
+        m += "0";
+    }
+
+    //length of message in binary, padded, and extended
+    //to a 64 bit representation
+    let ml = (message.length * CHAR_SIZE).toString(2);
+    ml = pad(ml, 8);
+    ml = "0".repeat(64 - ml.length) + ml;
+
+    return m + ml;
+}
+
+/**
+ * Hashes message using SHA-1 Cryptographic Hash Function
+ *
+ * @param {string} message - message to hash
+ * @return {string} - message digest (hash value)
+ */
+function SHA1(message) {
+    //main variables
+    let H0 = 0x67452301;
+    let H1 = 0xEFCDAB89;
+    let H2 = 0x98BADCFE;
+    let H3 = 0x10325476;
+    let H4 = 0xC3D2E1F0;
+
+    //pre-process message and split into 512 bit chunks
+    let bits = preProcess(message);
+    let chunks = chunkify(bits, 512);
+    
+    chunks.forEach(function(chunk, i) {
+        //break each chunk into 16 32-bit words
+        let words = chunkify(chunk, 32);
+
+        //extend 16 32-bit words to 80 32-bit words
+        for (let i = 16; i < 80; i++) {
+            let val = [words[i - 3], words[i - 8], words[i - 14], words[i - 16]]
+                .map(e => parseInt(e, 2))
+                .map(e => e >>> 0)
```

Comment:
Could you please explain why do we need `.map(e => e >>> 0)` ?  Doesn't e already fits into 32 bit?

Reply:
Since the words in each chunk are stored as bitstring representations, they must be converted to integers using `parseInt`. As I understand it, integers in Javascript are signed by default, so before using the xor operator (`^`), they must be converted to unsigned ints using `>>> 0`.

Reply:
The reason why `e >>> 0` is not required here is the way `parseInt` function works. It doesn't view leading 1 or 0 as sign bit, instead it decides that number is negative if the string starts from minus '-' sign. Thus in our case `e` is always positive.

Reply:
Oh okay I see now, thank you for the explanation. I'll remove the line and commit, as well as resolve the comment below.

Reply:
I'm sorry, this is my first pull request. I'm just wondering if I would now merge this PR or whether someone else with merge permissions would. Sorry for the confusion.

Reply:
Sorry for the delay, I will merge it right now :)